packages/charts/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx

```diff
@@ -1251,58 +1264,85 @@ export class VerticalBarChartBase
-        let reqWidth = (this._xAxisLabels.length + (this._xAxisLabels.length - 1) * barGapRate) * barWidth;
+        const bandwidth = totalWidth / numBars;
+        const barWidth = getBarWidth(this.props.barWidth, this.props.maxBarWidth, bandwidth, this.props.mode);
+        let reqWidth = numBars * barWidth;
         const margin1 = (totalWidth - reqWidth) / 2;
 
-        // Calculate the remaining width after accounting for the space required to render x-axis labels
-        const step = calculateLongestLabelWidth(this._xAxisLabels) + 20;
-        reqWidth = (this._xAxisLabels.length - this._xAxisInnerPadding) * step;
-        const margin2 = (totalWidth - reqWidth) / 2;
+        let margin2 = Number.POSITIVE_INFINITY;
+        if (!this.props.hideTickOverlap) {
+          // Calculate the remaining width after accounting for the space required to render x-axis labels
+          const step = calculateLongestLabelWidth(uniqueX as string[]) + 20;
+          reqWidth = (uniqueX.length - this._xAxisInnerPadding) * step;
+          margin2 = (totalWidth - reqWidth) / 2;
+        }
 
         this._domainMargin = MIN_DOMAIN_MARGIN + Math.max(0, Math.min(margin1, margin2));
       }
     } else {
-      const uniqueX: Record<number, number | Date> = {};
-      this.props.data?.forEach(point => {
-        if (point.x instanceof Date) {
-          uniqueX[point.x.getTime()] = point.x;
-        } else {
-          uniqueX[point.x as number] = point.x as number;
-        }
-      });
-      const data = Object.values(uniqueX) as number[] | Date[];
+      if (this.props.mode === 'histogram') {
+        // Try center-aligning the bars to eliminate any gaps caused by a restricted barWidth.
+        // This only works if the bin centers are consistent across all legend groups; otherwise,
+        // the calculated domainMargin may be too small.
+        const barWidth = this.props.maxBarWidth!;
+        const reqWidth = numBars * barWidth;
+        this._domainMargin += Math.max(0, (totalWidth - reqWidth) / 2);
+      }
+
+      // The histogram may appear distorted when bin centers/sizes vary across different legend groups.
+      // Currently, we calculate the appropriate bar width using the closest unique x-values to make
+      // the bars of the same legend group adjacent. But these x-values can come from different legend groups
+      // and result in misleading visuals. Even if we compute bar widths separately within each group,
+      // we still lack support for rendering bars with different widths and must use the minimum width,
+      // which can cause the same issue.
+      // Solution: Instead of estimating the appropriate bar width, render each bar to span the full range
+      // of its corresponding bin explicitly.
       this._barWidth = getBarWidth(
         this.props.barWidth,
         this.props.maxBarWidth,
-        calculateAppropriateBarWidth(data, totalWidth),
+        calculateAppropriateBarWidth(
```

Comment:
Followup. Add test schemas in the test app for all cases you have tested here.

Reply:
Ok